README.md
# cartograf/geocoder

[![Latest Version on Packagist][ico-version]][link-packagist]
[![Software License][ico-license]](LICENSE.md)
[![Total Downloads][ico-downloads]][link-downloads]

This is a bare-bones geocoding framework using cUrl to make http requests. I wrote it for myself as a utility class, mainly for
geocoding from the command line.

It's intended to be easily extensible for custom geocoding APIs and to get out of the way as much as possible.
`willdurand/geocoder` might be a better framework for you if you want something more robust or complex.

## Install

Via Composer

``` bash
$ composer require cartograf/geocoder
```

## Usage

You'll need a class which extends the abstract `Geocoder` base class and defines the following methods:

* `getQueryString(Address $address)` converts an `Address` into the portion of the query string which follows the `baseUrl`. Note that for now this needs to specify JSON output as well.
* `getBaseUrl()` - defaults to `$this->baseUrl` if that is set via `setBaseUrl()`
* `getLatLngFromResult(\stdClass $result)` consumes the json-decoded result object and returns an object with `latitude` and `longitude` properties. Note that this is the final return value so you could add other properties as well if you need them.
 
Classes implementing MapQuest Nominatim (just set API key) and geocoding via an AWS EC2 instance of [Geolytica's geocoder](https://aws.amazon.com/marketplace/pp/B013CW6HOA) are included.

``` php
$ch = curl_init() or die("curl_init failed");
$geocoder = new Cartograf\Geocoder($ch);
echo $geocoder->geoCode($address);
curl_close($ch);
```

## Change log

Please see [CHANGELOG](CHANGELOG.md) for more information what has changed recently.

## Testing

Not yet implemented.

``` bash
$ composer test
```

## Contributing

Feel free to extend this and submit PRs! I welcome contributions.

## Credits

- [Tim Stallmann][link-author]

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.

[ico-version]: https://img.shields.io/packagist/v/cartograf/geocoder.svg?style=flat-square
[ico-license]: https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square
[ico-travis]: https://img.shields.io/travis/cartograf/geocoder/master.svg?style=flat-square
[ico-scrutinizer]: https://img.shields.io/scrutinizer/coverage/g/cartograf/geocoder.svg?style=flat-square
[ico-code-quality]: https://img.shields.io/scrutinizer/g/cartograf/geocoder.svg?style=flat-square
[ico-downloads]: https://img.shields.io/packagist/dt/cartograf/geocoder.svg?style=flat-square

[link-packagist]: https://packagist.org/packages/cartograf/geocoder
[link-travis]: https://travis-ci.org/cartograf/geocoder
[link-scrutinizer]: https://scrutinizer-ci.com/g/cartograf/geocoder/code-structure
[link-code-quality]: https://scrutinizer-ci.com/g/cartograf/geocoder
[link-downloads]: https://packagist.org/packages/cartograf/geocoder
[link-author]: https://github.com/timstallmann
[link-contributors]: ../../contributors


CHANGELOG.md
# Changelog

All Notable changes to `geocoder` will be documented in this file.

Updates should follow the [Keep a CHANGELOG](http://keepachangelog.com/) principles.

## Release 1.0.2 - 2017-10-05

### Added
- US Census Geocoder plugin
- Google Geocoder plugin

### Added
- Nothing

### Deprecated
- Nothing

### Fixed
- Nothing

### Removed
- Nothing

### Security
- Nothing


src/AWS.php
<?php
/**
 * Created by PhpStorm.
 * User: tim
 * Date: 2/7/16
 * Time: 9:13 AM
 */

namespace Cartograf\GeoCoder;

use Cartograf\GeoCoder\Geocoder;

class AWS extends GeoCoder {
  protected function getQueryStringFromRawText($location) {
    // TODO: Implement getQueryStringFromRawText() method.
  }

  protected function getQueryString(Address $address) {
    $full_address = rawurlencode($address->street) . "," . rawurlencode($address->city) . "," . rawurlencode($address->state) . "," . rawurlencode($address->postalCode) . "," . rawurlencode($address->country);
    return $full_address . '?json=1';
  }

  protected function getLatLngFromResult(\stdClass $result) {
    if ($result && property_exists($result, 'standard') && $result->standard->confidence >= $this->threshold) {
      $ret = new LatLng;
      $ret->longitude = $result->longt;
      $ret->latitude = $result->latt;
      return $ret;
    }
    else {
      return FALSE;
    }
  }
}

src/Google.php
<?php
/**
 * Created by PhpStorm.
 * User: tim
 * Date: 2/7/16
 * Time: 9:13 AM
 */

namespace Cartograf\GeoCoder;

/**
 * Class Google
 * @package Cartograf\GeoCoder
 *
 * Extension of the base GeoCoder class for use with Google Maps API.
 */
class Google extends GeoCoder {
  protected $apiKey;

  // Array of valid location types.
  protected $validLocationTypes;

  public function __construct($ch) {
    parent::__construct($ch);
    $this->validLocationTypes = array('street_address', 'route', 'intersection', 'premise', 'subpremise');
  }

  public function getBaseUrl() {
    if ($this->apiKey) {
      return 'https://maps.googleapis.com/maps/api/geocode/json?key=' . $this->apiKey;
    }
    else {
      return FALSE;
    }
  }

  protected function getQueryStringFromRawText($location) {
    if ($location !== FALSE && $location !== '') {
      return '&address=' . rawurlencode($location);
    }
    else {
      return FALSE;
    }
  }

  protected function getQueryString(Address $address) {
    $base = $this->getQueryStringFromRawText(sprintf("%s, %s, %s %s", $address->street, $address->city, $address->state, $address->postalCode));
    $return = '';
    if ($address->country) {
      $return .= '&components=country:' . rawurlencode($address->country);
    }
    if ($base) {
      return $base . $return;
    }
    return FALSE;
  }

  protected function getLatLngFromResult(\stdClass $result) {
    $return = FALSE;
    if ($result) {
      if (property_exists($result, "status") && $result->status == "OK") {
        if (count(array_intersect($result->results[0]->types, $this->validLocationTypes)) > 0) {
          $return = new LatLng;
          $return->latitude = $result->results[0]->geometry->location->lat;
          $return->longitude = $result->results[0]->geometry->location->lng;
        }
      }
    }

    return $return;
  }
}


src/Geocoder.php
<?php
/**
 * Created by PhpStorm.
 * User: tim
 * Date: 2/7/16
 * Time: 9:11 AM
 */

namespace Cartograf\GeoCoder;

abstract class GeoCoder {
  protected $curlHandle;
  protected $apiKey = FALSE;
  protected $baseUrl = '';
  public $threshold = 0.5;

  abstract protected function getQueryString(Address $address);
  abstract protected function getQueryStringFromRawText($location);
  abstract protected function getLatLngFromResult(\stdClass $result);

  public function __construct($ch) {
    $this->curlHandle = $ch;
  }

  public function getBaseUrl() {
    if ($this->baseUrl <> '') {
      return $this->baseUrl;
    }
    else {
      return FALSE;
    }
  }

  public function setBaseUrl($url) {
    $this->baseUrl = $url;
  }

  public function setApiKey($api_key) {
    $this->apiKey = $api_key;
  }

  /**
   * Geocode a given address.
   *
   * @param \Cartograf\GeoCoder\Address $address
   *   Address object with fields specified, or NULL if using location.
   * @param null $location
   *   Raw string location with no fields specified.
   *
   * @return bool
   */
  public function geoCode(Address $address = NULL, $location = NULL) {
    if ($address !== NULL) {
      $query_string = $this->getQueryString($address);
    }
    else {
      $query_string = $this->getQueryStringFromRawText($location);
    }
    if ($this->getBaseUrl() && $query_string) {
      $url = $this->getBaseUrl() . $query_string;
      curl_setopt($this->curlHandle, CURLOPT_URL, $url);
      curl_setopt($this->curlHandle, CURLOPT_RETURNTRANSFER, TRUE);
      curl_setopt($this->curlHandle, CURLOPT_SSL_VERIFYPEER, FALSE);

      $result = json_decode(curl_exec($this->curlHandle));

      if ($result) {
          $latLng = $this->getLatLngFromResult($result);
          if ($latLng === FALSE) {
              throw new \Exception('Could not fetch latitude/longitude from result: ' . print_r($result, TRUE));
          }
          return $latLng;
      }
      else {
          throw new \Exception(curl_error($this->curlHandle));
      }
    }
    else {
      return FALSE;
    }
  }
}


src/MapQuestNominatim.php
<?php
/**
 * Created by PhpStorm.
 * User: tim
 * Date: 2/7/16
 * Time: 9:13 AM
 */

namespace Cartograf\GeoCoder;

class MapQuestNominatim extends GeoCoder {
  protected $apiKey;

  // Ranked array of the possibly values of the first 2 characters of
  // geocodeQualityCode from result, ordered from most precise to least.
  protected $granularityRanking;

  // Minimum precision which will be accepted.
  protected $minGranularity;

  public function __construct($ch) {
    parent::__construct($ch);
    $this->granularityRanking = array_flip(array('P1', 'L1', 'I1', 'B1', 'B2', 'B3', 'Z4', 'Z3', 'Z2', 'A6', 'Z1', 'A5', 'A4', 'A3', 'A2', 'A1'));
    $this->minGranularity = 'A6';
  }

  public function setMinGranularity($min_granularity) {
    $this->minGranularity = $min_granularity;
  }

  public function getBaseUrl() {
    if ($this->apiKey) {
      return 'http://open.mapquestapi.com/geocoding/v1/address?key=' . $this->apiKey;
    }
    else {
      return FALSE;
    }
  }

  protected function getQueryStringFromRawText($location) {
    $base = '&outFormat=json&thumbMaps=false&maxResults=1';
    if ($location !== FALSE && $location !== '') {
      return $base . '&location=' . rawurlencode($location);
    }
    else {
      return FALSE;
    }
  }

  protected function getQueryString(Address $address) {
    $base = '&outFormat=json&thumbMaps=false&maxResults=1';
    $return = '';
    if ($address->street) {
      $return .= '&street=' . rawurlencode($address->street);
    }
    if ($address->city) {
      $return .= '&city=' . rawurlencode($address->city);
    }
    if ($address->state) {
      $return .= '&state=' . rawurlencode($address->state);
    }
    if ($address->postalCode) {
      $return .= '&postalCode=' . rawurlencode($address->postalCode);
    }
    if ($address->country) {
      $return .= '&country=' . rawurlencode($address->country);
    }
    if ($return <> '') {
      return $base . $return;
    }
    return FALSE;
  }

  protected function getLatLngFromResult(\stdClass $result) {
    $return = FALSE;
    if ($result) {
      if (property_exists($result,
          "results") && is_array($result->results) && property_exists($result->results[0],
          "locations") && is_array($result->results[0]->locations)
      ) {
        if (count($result->results[0]->locations) > 0) {
          foreach ($result->results[0]->locations as $location) {
            if ($this->granularityRanking[substr($location->geocodeQualityCode,
                0,
                2)] <= $this->granularityRanking[$this->minGranularity]
            ) {
              $return = new LatLng;
              $return->latitude = $result->results[0]->locations[0]->latLng->lat;
              $return->longitude = $result->results[0]->locations[0]->latLng->lng;
              break;
            }
          }
        }
      }
      elseif (!empty($result->lat) && !empty($result->lng)) {
        $return = new LatLng;
        $return->latitude = $result->lat;
        $return->longitude = $result->lng;
      }
    }

    return $return;
  }
}


src/USCensus.php
<?php
/**
 * Created by PhpStorm.
 * User: tim
 * Date: 2/7/16
 * Time: 9:13 AM
 */

namespace Cartograf\GeoCoder;

/**
 * Class USCensus
 * @package Cartograf\GeoCoder
 *
 * Extension of the base GeoCoder class for use with US Census Geocoding API.
 * More info at https://geocoding.geo.census.gov/geocoder
 */
class USCensus extends GeoCoder {

  // Benchmark dataset to use for geocoding.
  protected $benchmark;

  public function __construct($ch) {
    parent::__construct($ch);
    $this->benchmark = 'Public_AR_Current';
  }

  public function setBenchmark($benchmark) {
    $this->benchmark = $benchmark;
  }

  public function getBaseUrl() {
    return 'https://geocoding.geo.census.gov/geocoder/locations/';
  }

  protected function getQueryStringFromRawText($location) {
    if ($location !== FALSE && $location !== '') {
      return sprintf('onelineaddress?format=json&benchmark=%s&address=%s', $this->benchmark, rawurlencode($location));
    }
    else {
      return FALSE;
    }
  }

  protected function getQueryString(Address $address) {
    $base = sprintf('address?format=json&benchmark=%s', $this->benchmark);
    $return = '';
    if ($address->street) {
      $return .= '&street=' . rawurlencode($address->street);
    }
    if ($address->city) {
      $return .= '&city=' . rawurlencode($address->city);
    }
    if ($address->state) {
      $return .= '&state=' . rawurlencode($address->state);
    }
    if ($address->postalCode) {
      $return .= '&zip=' . rawurlencode($address->postalCode);
    }
    if ($return <> '') {
      return $base . $return;
    }
    return FALSE;
  }

  protected function getLatLngFromResult(\stdClass $result) {
    $return = FALSE;
    if ($result && property_exists($result, "result")) {
      $result = $result->result;
      if (property_exists($result, "addressMatches") && count($result->addressMatches) > 0) {
        $return = new LatLng;
        $return->latitude = $result->addressMatches[0]->coordinates->y;
        $return->longitude = $result->addressMatches[0]->coordinates->x;
      }
    }

    return $return;
  }
}
